File: backend/app/api/v1/endpoints/application_fields.py
"""
Application field configuration API endpoints
"""

import io
import logging
from datetime import datetime, timezone
from typing import Any, Dict, List

from fastapi import APIRouter, Depends, File, HTTPException, Path, UploadFile, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.cache import cached, invalidate
from app.core.deps import get_db
from app.core.path_security import validate_upload_file
from app.core.security import get_current_user, require_admin
from app.models.application_field import ApplicationDocument, ApplicationField, FieldType
from app.models.user import User
from app.schemas.application_field import (
    ApplicationDocumentCreate,
    ApplicationDocumentUpdate,
    ApplicationFieldCreate,
    ApplicationFieldUpdate,
)
from app.schemas.response import ApiResponse
from app.services.application_field_service import ApplicationFieldService

router = APIRouter()
logger = logging.getLogger(__name__)


# Application Field endpoints
@router.get("/fields/{scholarship_type}")
async def get_fields_by_scholarship_type(
    scholarship_type: str, db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)
):
    """Get all fields for a scholarship type"""
    fields = await _get_fields_by_scholarship_type_cached(scholarship_type, db)
    return ApiResponse(success=True, message=f"Fields retrieved for {scholarship_type}", data=fields)


@cached(
    key_fn=lambda scholarship_type, *_, **__: f"fields:{scholarship_type}",
    ttl=86400,  # 24 h
)
async def _get_fields_by_scholarship_type_cached(scholarship_type: str, db: AsyncSession):
    service = ApplicationFieldService(db)
    return await service.get_fields_by_scholarship_type(scholarship_type)


@router.post("/fields")
async def create_field(
    field_data: ApplicationFieldCreate, db: AsyncSession = Depends(get_db), current_user: User = Depends(require_admin)
):
    """Create a new application field"""
    service = ApplicationFieldService(db)
    field = await service.create_field(field_data, current_user.id)
    await invalidate(f"fields:{field_data.scholarship_type}")
    await invalidate(f"formconfig:{field_data.scholarship_type}")
    return ApiResponse(success=True, message="Application field created successfully", data=field)


@router.put("/fields/{field_id}")
async def update_field(
    field_id: int,
    field_data: ApplicationFieldUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Update an application field"""
    service = ApplicationFieldService(db)

    # Pre-fetch the existing row so we can re-check the export flag against
    # the post-merge field_type. The validator on ApplicationFieldUpdate can't
    # see the existing DB value when field_type is omitted from the payload.
    existing_query = select(ApplicationField).where(ApplicationField.id == field_id)
    existing_result = await db.execute(existing_query)
    existing_field = existing_result.scalar_one_or_none()

    if not existing_field:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application field not found")

    update_payload = field_data.model_dump(exclude_unset=True)
    merged_field_type = update_payload.get("field_type", existing_field.field_type)
    merged_include_flag = update_payload.get(
        "include_in_college_export",
        existing_field.include_in_college_export,
    )

    # Re-check export flag against the merged field_type (validator on the
    # update schema can't see the existing DB value when field_type is omitted).
    if merged_include_flag and merged_field_type != FieldType.TEXT.value:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="include_in_college_export 僅支援 field_type='text'",
        )

    field = await service.update_field(field_id, field_data, current_user.id)

    if not field:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application field not found")

    # `field` is a Pydantic response; pull scholarship_type either from response
    # or fall back to a broad invalidate.
    sct = getattr(field, "scholarship_type", None)
    if sct:
        await invalidate(f"fields:{sct}")
        await invalidate(f"formconfig:{sct}")
    else:
        await invalidate("fields:")
        await invalidate("formconfig:")

    return ApiResponse(success=True, message="Application field updated successfully", data=field)


@router.delete("/fields/{field_id}")
async def delete_field(field_id: int, db: AsyncSession = Depends(get_db), current_user: User = Depends(require_admin)):
    """Delete an application field"""
    service = ApplicationFieldService(db)
    success = await service.delete_field(field_id)

    if not success:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application field not found")

    # We don't have scholarship_type cheaply post-delete; clobber the namespace.
    await invalidate("fields:")
    await invalidate("formconfig:")

    return ApiResponse(success=True, message="Application field deleted successfully", data=True)


# Application Document endpoints
@router.get("/documents/{scholarship_type}")
async def get_documents_by_scholarship_type(
    scholarship_type: str, db: AsyncSession = Depends(get_db), current_user: User = Depends(get_current_user)
):
    """Get all documents for a scholarship type"""
    documents = await _get_documents_by_scholarship_type_cached(scholarship_type, db)
    return ApiResponse(success=True, message=f"Documents retrieved for {scholarship_type}", data=documents)


@cached(
    key_fn=lambda scholarship_type, *_, **__: f"documents:{scholarship_type}",
    ttl=86400,  # 24 h
)
async def _get_documents_by_scholarship_type_cached(scholarship_type: str, db: AsyncSession):
    service = ApplicationFieldService(db)
    return await service.get_documents_by_scholarship_type(scholarship_type)


@router.post("/documents")
async def create_document(
    document_data: ApplicationDocumentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Create a new application document"""
    service = ApplicationFieldService(db)
    document = await service.create_document(document_data, current_user.id)
    await invalidate(f"documents:{document_data.scholarship_type}")
    await invalidate(f"formconfig:{document_data.scholarship_type}")
    return ApiResponse(success=True, message="Application document created successfully", data=document)


@router.put("/documents/{document_id}")
async def update_document(
    document_id: int,
    document_data: ApplicationDocumentUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Update an application document"""
    service = ApplicationFieldService(db)
    document = await service.update_document(document_id, document_data, current_user.id)

    if not document:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application document not found")

    sct = getattr(document, "scholarship_type", None)
    if sct:
        await invalidate(f"documents:{sct}")
        await invalidate(f"formconfig:{sct}")
    else:
        await invalidate("documents:")
        await invalidate("formconfig:")

    return ApiResponse(success=True, message="Application document updated successfully", data=document)


@router.delete("/documents/{document_id}")
async def delete_document(
    document_id: int, db: AsyncSession = Depends(get_db), current_user: User = Depends(require_admin)
):
    """Delete an application document"""
    service = ApplicationFieldService(db)
    success = await service.delete_document(document_id)

    if not success:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application document not found")

    await invalidate("documents:")
    await invalidate("formconfig:")

    return ApiResponse(success=True, message="Application document deleted successfully", data=True)


# Combined form configuration endpoints
# SECURITY: This endpoint serves database-driven JSON configuration only.
# No file system access or source code exposure. Path parameter validated
# and used only in parameterized SQL queries.
@router.get("/form-config/{scholarship_type}")
async def get_scholarship_form_config(
    scholarship_type: str = Path(..., pattern=r"^[a-z_]{1,50}$"),
    include_inactive: bool = False,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get complete form configuration for a scholarship type"""
    try:
        logger.debug(f"API: Getting form config for {scholarship_type}")

        # 管理員可以看到所有欄位（包括停用的），其他用戶只能看到啟用的
        is_admin = current_user.role in ["admin", "super_admin"]
        should_include_inactive = include_inactive or is_admin

        # We key cache on the *resolved* should_include_inactive so admin and
        # non-admin views never share an entry. user_id is intentionally NOT
        # part of the key — the config is identical for everyone in the same
        # admin/non-admin bucket; user_id is only forwarded for audit logging.
        config = await _get_scholarship_form_config_cached(
            scholarship_type, should_include_inactive, db, current_user.id
        )

        logger.info(f"API: Form config retrieved successfully for {scholarship_type}")
        return ApiResponse(success=True, message=f"Form configuration retrieved for {scholarship_type}", data=config)
    except Exception as e:
        logger.exception(f"API: Error getting form config for {scholarship_type}")
        # Return 404 for invalid scholarship types (proper HTTP semantics)
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Form configuration not found for scholarship type: {scholarship_type}",
        ) from e


@cached(
    key_fn=lambda scholarship_type, should_include_inactive, *_, **__: (
        f"formconfig:{scholarship_type}:{'all' if should_include_inactive else 'active'}"
    ),
    ttl=21600,  # 6 h
)
async def _get_scholarship_form_config_cached(
    scholarship_type: str, should_include_inactive: bool, db: AsyncSession, user_id: int
):
    service = ApplicationFieldService(db)
    return await service.get_scholarship_form_config(scholarship_type, should_include_inactive, user_id=user_id)


class FormConfigSaveRequest(BaseModel):
    """Schema for saving form configuration"""

    fields: List[Dict[str, Any]]
    documents: List[Dict[str, Any]]


@router.post("/form-config/{scholarship_type}")
async def save_scholarship_form_config(
    scholarship_type: str,
    config_data: FormConfigSaveRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Save complete form configuration for a scholarship type"""
    service = ApplicationFieldService(db)

    config = await service.save_scholarship_form_config(
        scholarship_type=scholarship_type,
        fields_data=config_data.fields,
        documents_data=config_data.documents,
        user_id=current_user.id,
    )

    await invalidate(f"fields:{scholarship_type}")
    await invalidate(f"documents:{scholarship_type}")
    await invalidate(f"formconfig:{scholarship_type}")

    return ApiResponse(success=True, message=f"Form configuration saved for {scholarship_type}", data=config)


# Example file management endpoints
@router.post("/documents/{document_id}/upload-example")
async def upload_document_example(
    document_id: int,
    file: UploadFile = File(...),
    current_user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
):
    """
    Upload example file for an application document

    Args:
        document_id: Application document ID
        file: Example file to upload
        current_user: Current authenticated admin user
        db: Database session

    Returns:
        ApiResponse with uploaded file object name
    """
    # SECURITY: Comprehensive file validation (CLAUDE.md triple validation)
    # Read file first for size validation
    file_content = await file.read()
    file_size = len(file_content)

    validate_upload_file(
        filename=file.filename,
        allowed_extensions=[".pdf", ".jpg", ".jpeg", ".png", ".doc", ".docx"],
        max_size_mb=10,
        file_size=file_size,
        allow_unicode=True,
    )

    # Get document from database
    stmt = select(ApplicationDocument).where(ApplicationDocument.id == document_id)
    result = await db.execute(stmt)
    document = result.scalar_one_or_none()

    if not document:
        raise HTTPException(status_code=404, detail="Application document not found")

    try:
        from app.services.minio_service import minio_service

        # Generate object name with timestamp
        timestamp = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
        file_extension = file.filename.rsplit(".", 1)[-1] if "." in file.filename else "pdf"
        object_name = f"examples/document_{document_id}_{timestamp}.{file_extension}"

        # Note: file_content already read above for validation

        # Upload to MinIO
        minio_service.client.put_object(
            bucket_name=minio_service.default_bucket,
            object_name=object_name,
            data=io.BytesIO(file_content),
            length=len(file_content),
            content_type=file.content_type or "application/octet-stream",
        )

        # Delete old example file if exists
        if document.example_file_url:
            try:
                minio_service.client.remove_object(
                    bucket_name=minio_service.default_bucket, object_name=document.example_file_url
                )
            except Exception:
                logger.warning("Failed to delete old example file", exc_info=True)

        # Update database with new object name
        document.example_file_url = object_name
        document.updated_by = current_user.id
        await db.commit()

        return ApiResponse(success=True, message="範例文件上傳成功", data={"example_file_url": object_name})

    except Exception as e:
        logger.exception("Failed to upload example file")
        await db.rollback()
        raise HTTPException(status_code=500, detail="範例文件上傳失敗") from e


@router.get("/documents/{document_id}/example")
async def get_document_example(
    document_id: int, current_user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
):
    """
    Get example file for an application document and proxy download from MinIO

    Args:
        document_id: Application document ID
        current_user: Current authenticated user
        db: Database session

    Returns:
        Streaming response with file content
    """
    # Get document from database
    stmt = select(ApplicationDocument).where(ApplicationDocument.id == document_id)
    result = await db.execute(stmt)
    document = result.scalar_one_or_none()

    if not document or not document.example_file_url:
        raise HTTPException(status_code=404, detail="範例文件不存在")

    try:
        from app.services.minio_service import minio_service

        # Download from MinIO
        response = minio_service.client.get_object(
            bucket_name=minio_service.default_bucket, object_name=document.example_file_url
        )

        file_content = response.read()

        # Determine content type based on file extension
        file_extension = document.example_file_url.rsplit(".", 1)[-1].lower()
        content_type_map = {
            "pdf": "application/pdf",
            "doc": "application/msword",
            "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "jpg": "image/jpeg",
            "jpeg": "image/jpeg",
            "png": "image/png",
        }
        content_type = content_type_map.get(file_extension, "application/octet-stream")

        # Return as streaming response
        return StreamingResponse(
            io.BytesIO(file_content),
            media_type=content_type,
            headers={
                "Content-Disposition": f"inline; filename*=UTF-8''{document.document_name}_example.{file_extension}"
            },
        )

    except Exception as e:
        logger.exception("Failed to get example file")
        raise HTTPException(status_code=500, detail="範例文件讀取失敗") from e


@router.delete("/documents/{document_id}/example")
async def delete_document_example(
    document_id: int, current_user: User = Depends(require_admin), db: AsyncSession = Depends(get_db)
):
    """
    Delete example file for an application document

    Args:
        document_id: Application document ID
        current_user: Current authenticated admin user
        db: Database session

    Returns:
        ApiResponse with success status
    """
    # Get document from database
    stmt = select(ApplicationDocument).where(ApplicationDocument.id == document_id)
    result = await db.execute(stmt)
    document = result.scalar_one_or_none()

    if not document:
        raise HTTPException(status_code=404, detail="Application document not found")

    if not document.example_file_url:
        raise HTTPException(status_code=404, detail="範例文件不存在")

    try:
        from app.services.minio_service import minio_service

        # Delete from MinIO
        minio_service.client.remove_object(
            bucket_name=minio_service.default_bucket, object_name=document.example_file_url
        )

        # Clear database field
        document.example_file_url = None
        document.updated_by = current_user.id
        await db.commit()

        return ApiResponse(success=True, message="範例文件刪除成功", data=True)

    except Exception as e:
        logger.exception("Failed to delete example file")
        await db.rollback()
        raise HTTPException(status_code=500, detail="範例文件刪除失敗") from e

File: frontend/lib/api/modules/application-fields.ts
/**
 * Application Fields API Module (OpenAPI-typed)
 *
 * Manages dynamic form fields and documents for scholarship applications:
 * - Form configuration
 * - Custom field definitions
 * - Document requirements
 * - Example file uploads
 *
 * Now using openapi-fetch for full type safety from backend OpenAPI schema
 */

import { typedClient } from '../typed-client';
import { toApiResponse } from '../compat';
import type { ApiResponse } from '../../api.legacy';

type ScholarshipFormConfig = {
  scholarship_type: string;
  fields: any[];
  documents: any[];
};

type FormConfigSaveRequest = {
  fields: any[];
  documents: any[];
};

type ApplicationField = {
  id: number;
  scholarship_type: string;
  field_name: string;
  field_label: string;
  field_type: string;
  required: boolean;
  display_order: number;
  is_active: boolean;
  options?: any;
};

type ApplicationFieldCreate = Omit<ApplicationField, 'id'>;
type ApplicationFieldUpdate = Partial<ApplicationFieldCreate>;

type ApplicationDocument = {
  id: number;
  scholarship_type: string;
  document_name: string;
  document_label: string;
  required: boolean;
  display_order: number;
  is_active: boolean;
  example_file_path?: string;
  description?: string;
};

type ApplicationDocumentCreate = Omit<ApplicationDocument, 'id'>;
type ApplicationDocumentUpdate = Partial<ApplicationDocumentCreate>;

export function createApplicationFieldsApi() {
  return {
    /**
     * Get form configuration for a scholarship type
     * Type-safe: Path parameter and query parameters validated against OpenAPI
     */
    getFormConfig: async (
      scholarshipType: string,
      includeInactive: boolean = false
    ): Promise<ApiResponse<ScholarshipFormConfig>> => {
      const response = await typedClient.raw.GET('/api/v1/application-fields/form-config/{scholarship_type}', {
        params: {
          path: { scholarship_type: scholarshipType },
          query: { include_inactive: includeInactive },
        },
      });
      return toApiResponse<ScholarshipFormConfig>(response);
    },

    /**
     * Save form configuration for a scholarship type
     * Type-safe: Path parameter and request body validated against OpenAPI
     */
    saveFormConfig: async (
      scholarshipType: string,
      config: FormConfigSaveRequest
    ): Promise<ApiResponse<ScholarshipFormConfig>> => {
      const response = await typedClient.raw.POST('/api/v1/application-fields/form-config/{scholarship_type}', {
        params: { path: { scholarship_type: scholarshipType } },
        body: config,
      });
      return toApiResponse<ScholarshipFormConfig>(response);
    },

    /**
     * Get all fields for a scholarship type
     * Type-safe: Path parameter validated against OpenAPI
     */
    getFields: async (
      scholarshipType: string
    ): Promise<ApiResponse<ApplicationField[]>> => {
      const response = await typedClient.raw.GET('/api/v1/application-fields/fields/{scholarship_type}', {
        params: { path: { scholarship_type: scholarshipType } },
      });
      return toApiResponse<ApplicationField[]>(response);
    },

    /**
     * Create a new field
     * Type-safe: Request body validated against OpenAPI
     */
    createField: async (
      fieldData: ApplicationFieldCreate
    ): Promise<ApiResponse<ApplicationField>> => {
      const response = await typedClient.raw.POST('/api/v1/application-fields/fields', {
        body: fieldData as any, // Frontend includes additional optional fields not in OpenAPI schema
      });
      return toApiResponse<ApplicationField>(response);
    },

    /**
     * Update an existing field
     * Type-safe: Path parameter and request body validated against OpenAPI
     */
    updateField: async (
      fieldId: number,
      fieldData: ApplicationFieldUpdate
    ): Promise<ApiResponse<ApplicationField>> => {
      const response = await typedClient.raw.PUT('/api/v1/application-fields/fields/{field_id}', {
        params: { path: { field_id: fieldId } },
        body: fieldData,
      });
      return toApiResponse<ApplicationField>(response);
    },

    /**
     * Delete a field
     * Type-safe: Path parameter validated against OpenAPI
     */
    deleteField: async (fieldId: number): Promise<ApiResponse<boolean>> => {
      const response = await typedClient.raw.DELETE('/api/v1/application-fields/fields/{field_id}', {
        params: { path: { field_id: fieldId } },
      });
      return toApiResponse<boolean>(response);
    },

    /**
     * Get all documents for a scholarship type
     * Type-safe: Path parameter validated against OpenAPI
     */
    getDocuments: async (
      scholarshipType: string
    ): Promise<ApiResponse<ApplicationDocument[]>> => {
      const response = await typedClient.raw.GET('/api/v1/application-fields/documents/{scholarship_type}', {
        params: { path: { scholarship_type: scholarshipType } },
      });
      return toApiResponse<ApplicationDocument[]>(response);
    },

    /**
     * Create a new document requirement
     * Type-safe: Request body validated against OpenAPI
     */
    createDocument: async (
      documentData: ApplicationDocumentCreate
    ): Promise<ApiResponse<ApplicationDocument>> => {
      const response = await typedClient.raw.POST('/api/v1/application-fields/documents', {
        body: documentData as any, // Frontend includes additional optional fields not in OpenAPI schema
      });
      return toApiResponse<ApplicationDocument>(response);
    },

    /**
     * Update an existing document requirement
     * Type-safe: Path parameter and request body validated against OpenAPI
     */
    updateDocument: async (
      documentId: number,
      documentData: ApplicationDocumentUpdate
    ): Promise<ApiResponse<ApplicationDocument>> => {
      const response = await typedClient.raw.PUT('/api/v1/application-fields/documents/{document_id}', {
        params: { path: { document_id: documentId } },
        body: documentData,
      });
      return toApiResponse<ApplicationDocument>(response);
    },

    /**
     * Delete a document requirement
     * Type-safe: Path parameter validated against OpenAPI
     */
    deleteDocument: async (documentId: number): Promise<ApiResponse<boolean>> => {
      const response = await typedClient.raw.DELETE('/api/v1/application-fields/documents/{document_id}', {
        params: { path: { document_id: documentId } },
      });
      return toApiResponse<boolean>(response);
    },

    /**
     * Upload example file for a document
     * Type-safe: FormData upload with path parameter
     */
    uploadDocumentExample: async (
      documentId: number,
      file: File
    ): Promise<any> => {
      const formData = new FormData();
      formData.append("file", file);

      const token = typedClient.getToken();
      const baseURL =
        typeof window !== "undefined"
          ? ""
          : process.env.INTERNAL_API_URL || "http://localhost:8000";

      const response = await fetch(
        `${baseURL}/api/v1/application-fields/documents/${documentId}/upload-example`,
        {
          method: "POST",
          body: formData,
          credentials: "include",
          headers: {
            ...(token ? { Authorization: `Bearer ${token}` } : {}),
          },
        }
      );

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.detail || "Failed to upload example file");
      }

      return response.json();
    },

    /**
     * Delete example file for a document
     * Type-safe: Path parameter validated against OpenAPI
     */
    deleteDocumentExample: async (
      documentId: number
    ): Promise<ApiResponse<boolean>> => {
      const response = await typedClient.raw.DELETE('/api/v1/application-fields/documents/{document_id}/example', {
        params: { path: { document_id: documentId } },
      });
      return toApiResponse<boolean>(response);
    },
  };
}
